sql/src/main/java/io/crate/planner/consumer/CrossJoinConsumer.java

```diff
@@ -214,58 +215,84 @@ private Symbol replaceFieldsWithInputColumns(Symbol symbol, List<Symbol> inputSy
         }
 
         private NestedLoop toNestedLoop(List<QueriedTableRelation> queriedTables, ConsumerContext context) {
-            Iterator<QueriedTableRelation> iterator = queriedTables.iterator();
+            // we must start from right to left, so reverse list of source relations
+            Iterator<QueriedTableRelation> iterator = Lists.reverse(queriedTables).iterator();
             NestedLoop nl = null;
-            PlannedAnalyzedRelation left;
-            PlannedAnalyzedRelation right;
+            List<Symbol> outputs = new ArrayList<>();
+            OrderBy orderBy = null;
+            PlannedAnalyzedRelation leftPlan;
+            PlannedAnalyzedRelation rightPlan;
 
             Set<String> localExecutionNodes = ImmutableSet.of(clusterService.localNode().id());
             UUID jobId = context.plannerContext().jobId();
 
             while (iterator.hasNext()) {
-                QueriedTableRelation next = iterator.next();
+                QueriedTableRelation rightRelation = iterator.next();
+                rightPlan = consumingPlanner.plan(rightRelation, context);
 
-                if (nl == null) {
-                    assert iterator.hasNext();
-                    QueriedTableRelation second = iterator.next();
+                assert rightPlan != null;
 
-                    left = consumingPlanner.plan(next, context);
-                    right = consumingPlanner.plan(second, context);
-                    assert left != null && right != null;
+                MergePhase rightMerge = mergePhase(
+                        context,
+                        localExecutionNodes,
+                        jobId,
+                        rightPlan.resultNode(),
+                        rightRelation.querySpec().outputs(),
+                        rightRelation.querySpec().orderBy()
+                );
 
-                    MergePhase leftMerge = mergePhase(
+                assert iterator.hasNext() || nl != null;
+
+                MergePhase leftMerge;
+                if (nl == null && iterator.hasNext()) {
```

Comment:
imho the second check (`iterator.hasNext`) is not really needed because of the assertion before hand... and I think there is also an assertion somewhere that there must be at least 2 tables